test/api.jl

```diff
@@ -115,5 +116,30 @@ recommendation = provide_recommendation(trial_vector)
 recommendation_multi = provide_recommendation(trial_vector_multihyperparams)
 @test recommendation_multi == Dict(:x => 3, :y => 7)
 
+# Test case reproducing the issue in “obs_memo” function from the higher level
+space = Dict(
+    :a => HP.Choice(:a, [8, 9, 10]),
+    :b => HP.Choice(:b, [1, 2, 3])
+    )
+
+Random.seed!(10)
```

Comment:
Please comment this -- you shouldn't normally set the seed to do tests (and instead do property testing on samples), but for this particular case its ok. The idea is to comment it so that it is something discouraged from copying.

If you can figure out a way to specify this test without setting the random seed, it would be better. Finally, I think we'd need to use stablerngs because otherwise i can imagine this fails on different julia versions (the RNG seed isn't stable across julia versions, which is what the stablerngs package version addresses)

Reply:
In fact I think you dont need the seed because the choices are different values in the two distributions you never expect values sampled for b to match values sampled for a

Reply:
As this is a test at a high level which not only includes the bug function "obs_memo", bug also includes the following function about the posterior inference, if it doesn't set the seed, even the "obs_memo" returns the same nid, the posterior inference would return different results sometimes due to its internal random returned values.

Reply:
Thats why I said to use property tests -- in this case you have two disjoint sets to sample from and its clear to me that for `:a` we should never sample from the parmater range for `:b' and vice versa, and that is the terms in which this test should be specified

Reply:
To be concrete: 

instead of saying sample 30 for `:a` != sample 30 for `:b`, a more appropriate test would say

samples 26:30 for `:a` come from the valid values for `:a`
samples 26:30 for `b` come from the valid values for `:b`

We dont need to set a random seed and this test is more robust.

Reply:
> Thats why I said to use property tests -- in this case you have two disjoint sets to sample from and its clear to me that for `:a` we should never sample from the parmater range for `:b` and vice versa, and that is the terms in which this test should be specified

Both ":a" and ":b" samples from ":a", they will never sample from ":b". Although both sets have different elements, they have the same number of choices (3 choices). The "vals" for both params are the same, which is the index (1,2,3) of both choices. Thus both params have the same set of "vals". As the posterior inference only check the history records of "vals[nid]" not "parameter[nid]", both sets are same as another example (hp.choice('a', [1, 2, 3]) and hp.choice('b', [1, 2, 3])).

Reply:
> To be concrete:
> 
> instead of saying sample 30 for `:a` != sample 30 for `:b`, a more appropriate test would say
> 
> samples 26:30 for `:a` come from the valid values for `:a` samples 26:30 for `:b` come from the valid values for `:b`
> 
> We dont need to set a random seed and this test is more robust.

This way would not show the bug because even after the bug is fixed, it is still that samples 21:30 for :a come from the valid values for :a and samples 21:30 for :b come from the valid values for :b. (the random iterations are the first 20 iterations)

We are comparing the "trial[30].vals" value, not the "trila[30].parameters" value, and the "trial[30].vals" should be the same for both params before the bug is fixed while they should be different after the bug is fixed. Why we don't use the values from the 21 iteration is because after the bug is fixed, optimisation can not find the best answer (different results) so fast at the 21 iteration while it can find the best answer at the finally iteration (30).

Reply:
> it is still that samples 21:30 for :a come from the valid values for :a and samples 21:30 for :b come from the valid values for :b

This can't be true and also `:a` at 30 != `:b` at 30 because the values for `:a` and `:b` are from disjoint sets

The reason why when the bug is present `:a` at 30 and `:b` at 30 are the same is precisely because the samples from `:b` draws from the posterior for `:a`

Reply:
basically the robust form of the test is this:

```julia
a_set = [8, 9, 10]
b_set = [1, 2, 3]
avals = getindex.(getfield.(trials, Ref(:vals)),  Ref(:a))
bvals = getindex.(getfield.(trials, Ref(:vals)),  Ref(:b))

@test all(in.(avals, Ref(a_set)))
@test all(in.(bvals, Ref(b_set)))
```

Reply:
if you write it like this, there's no need of the seed, because it relies on the property that A intersect B is empty

Reply:
I think the seed is necessary here. We could have a short meeting to clarify this issue, it will speed up the process, let me know when is good for you I will send a teams invite.

Reply:
Did you try as I wrote? It fails as expected when I wrote it locally.

There's no "issue" here except that I'm asking for a more robust test that doesn't rely on randomness, by writing it in terms of expected properties (which is that samples come from valid distributions).

If you write the test in terms of these expected properties (which I think we can agree its true that we expect the samples come from their corresponding distributions), theres suddenly no need of the seed.

Reply:
In fact even as written, theres no need of the seed entirely **because** the sets of values for the two distributions are disjoint